.github/workflows/build_and_test.yml

```diff
@@ -1,164 +1,215 @@
-      run: |
-        python -m pip install flake8
-        flake8 .
-
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip setuptools wheel
+        shell: bash
+
+      - name: Linter
+        run: |
+          python -m pip install flake8
+          flake8 .
+
+  # Segregate all OSes to handle calling the scripts differently.
+  # This is necessary because of the tests to piping, which require
+  # a TTY, which is differently set up in each OS runner.
   build_and_test_linux:
     runs-on: ${{ matrix.platform }}
     strategy:
       matrix:
         platform: [ubuntu-latest]
-        python-version: [2.7, 3.6, 3.7, 3.8, 3.9]
+        python-version: [2.7, 3.6, 3.7, 3.8, 3.9, "3.10"]
 
     steps:
-    - uses: actions/checkout@v2
-
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v2
-      with:
-        python-version: ${{ matrix.python-version }}
-
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip setuptools wheel bump2version twine
-      shell: bash
-
-    - name: Linter
-      run: |
-        python -m pip install flake8
-        flake8 .
-
-    - name: Build
-      run: |
-        python --version
-        python setup.py sdist bdist_wheel
-        twine check dist/*.whl
-        twine check dist/*.tar.gz
-        bump2version --dry-run --verbose --allow-dirty patch
-        bump2version --dry-run --verbose --allow-dirty minor
-        bump2version --dry-run --verbose --allow-dirty major
-
-    - name: Test
-      run: |
-        python -m pip install .
-        script -q -e -c "python setup.py test"
-
-  test_macos:
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip setuptools wheel
+          python -m pip install --upgrade bump2version twine
+          python -m pip install --upgrade coverage
+        shell: bash
+
+      - name: Build
+        run: |
+          python --version
+          python setup.py sdist bdist_wheel
+          twine check dist/*.whl
+          twine check dist/*.tar.gz
+          bump2version --dry-run --verbose --allow-dirty patch
+          bump2version --dry-run --verbose --allow-dirty minor
+          bump2version --dry-run --verbose --allow-dirty major
+
+      - name: Test
+        run: |
+          python -m pip install .
+          script -q -e -c "coverage run -p setup.py test"
+
+      # Store coverage data for later
+      # https://hynek.me/articles/ditch-codecov-python/
+      - name: Upload coverage data
```

Comment:
This upload is no longer needed since we dropped codecov. Or is it? :thinking:

Reply:
And does it need to do that combine for every OS and python version or only for all of them in the end?

Reply:
It needs to upload the artifact at the end of every run, then the combination needs to grab the artifacts and combine them in one step. I should use 'store' actually, although upload is correct as well. I thought this was in another place in the code when I suggested 'combine'. Basically, it's uploading from the runner to a centralized location that can persist and be accessed by other future runners.

Reply:
Okay, now you said it differently. If ignoring those tests makes the YAML collapse to a single matrix, then my vote. You are right there are some commands OS-specific for the different builds to workaround the TTY. Can we make it to run the TTY at least on ubuntu, or not even?

Reply:
Thanks for explaining :+1: